vm-virtio/src/transport/mmio.rs

```diff
@@ -354,6 +407,84 @@ impl Pausable for MmioDevice {
     }
 }
 
-impl Snapshottable for MmioDevice {}
+const VIRTIO_MMIO_DEV_SNAPSHOT_ID: &str = "virtio_mmio_device";
+impl Snapshottable for MmioDevice {
+    fn id(&self) -> String {
+        format!(
+            "{}-{}",
+            VIRTIO_MMIO_DEV_SNAPSHOT_ID,
+            VirtioDeviceType::from(self.device.lock().unwrap().device_type())
+        )
+    }
+
+    fn snapshot(&self) -> std::result::Result<Snapshot, MigratableError> {
+        let snapshot =
+            serde_json::to_vec(&self.state()).map_err(|e| MigratableError::Snapshot(e.into()))?;
+
+        let snapshot_id = self.id();
+        let mut virtio_mmio_dev_snapshot = Snapshot::new(&snapshot_id);
+        virtio_mmio_dev_snapshot.add_data_section(SnapshotDataSection {
+            id: format!("{}-section", snapshot_id),
+            snapshot,
+        });
+
+        Ok(virtio_mmio_dev_snapshot)
+    }
+
+    fn restore(&mut self, snapshot: Snapshot) -> std::result::Result<(), MigratableError> {
+        let snapshot_id = self.id();
+        if let Some(virtio_mmio_dev_section) = snapshot
+            .snapshot_data
+            .get(&format!("{}-section", snapshot_id))
+        {
+            let virtio_mmio_dev_state =
+                match serde_json::from_slice(&virtio_mmio_dev_section.snapshot) {
+                    Ok(state) => state,
+                    Err(error) => {
+                        return Err(MigratableError::Restore(anyhow!(
+                            "Could not deserialize VIRTIO_MMIO_DEVICE {}",
+                            error
+                        )))
+                    }
+                };
+
+            // First restore the status of the virtqueues.
+            self.set_state(&virtio_mmio_dev_state).map_err(|e| {
+                MigratableError::Restore(anyhow!(
+                    "Could not restore VIRTIO_MMIO_DEVICE state {:?}",
+                    e
+                ))
+            })?;
+
+            // Then we can activate the device, as we know at this point that
+            // the virtqueues are in the right state and the device is ready
+            // to be activated, which will spawn each virtio worker thread.
+            if self.device_activated && self.is_driver_ready() && self.are_queues_valid() {
+                if let Some(interrupt_cb) = self.interrupt_cb.take() {
+                    if self.mem.is_some() {
+                        let mem = self.mem.as_ref().unwrap().clone();
+                        self.device
+                            .lock()
+                            .unwrap()
+                            .activate(
+                                mem,
+                                interrupt_cb,
+                                self.queues.clone(),
+                                self.queue_evts.split_off(0),
+                            )
+                            .expect("Failed to activate device");
```

Comment:
nit: Can we replace this panic with an error?

Reply:
You mean `map_err()`, right?

Reply:
Done